android/MobiledgeXSDKDemo/app/src/main/java/com/mobiledgex/sdkdemo/MainActivity.java

```diff
@@ -1899,121 +1903,149 @@ public void onDestroy() {
+                for (int i = 0; i < waypointMarkers.size(); i++) {
+                    Marker marker = waypointMarkers.get(i);
+                    // Converting between com.google.maps.model.LatLng
+                    // and com.google.android.gms.maps.model.Latlng is a pain!
+                    waypoints[i] = new DirectionsApiRequest.Waypoint(new com.google.maps.model.LatLng(marker.getPosition().latitude, marker.getPosition().longitude));
+                }
+                try {
+                    DirectionsResult calculatedRoutes = null;
+                        calculatedRoutes = DirectionsApi.newRequest(context)
+                                .alternatives(false)
+                                .mode(TravelMode.DRIVING)
+                                .origin(new com.google.maps.model.LatLng(startLatLng.latitude, startLatLng.longitude))
+                                .waypoints(waypoints)
+                                .destination(new com.google.maps.model.LatLng(endLatLng.latitude, endLatLng.longitude))
+                                .await();
+
+                    if (calculatedRoutes == null || calculatedRoutes.routes.length == 0) {
+                        Log.w(TAG, "calculatedRoutes has no content");
+                        showError("No driving route found.");
+                        toggleFabPlayRoute(View.GONE);
+                        return;
+                    }
+                    Log.d(TAG, "calculatedRoutes=" + calculatedRoutes.routes.length);
+                    toggleFabPlayRoute(View.VISIBLE);
+
+                    int numPoints = 0;
+                    //Loop through legs and steps to get encoded polylines of each step
+                    for (DirectionsRoute route : calculatedRoutes.routes) {
+                        List<LatLng> path = new ArrayList<>();
+
+                        if (route.legs != null) {
+                            for (int i = 0; i < route.legs.length; i++) {
+                                DirectionsLeg leg = route.legs[i];
+                                if (leg.steps != null) {
+                                    for (int j = 0; j < leg.steps.length; j++) {
+                                        DirectionsStep step = leg.steps[j];
+                                        if (step.steps != null && step.steps.length > 0) {
+                                            for (int k = 0; k < step.steps.length; k++) {
+                                                DirectionsStep step1 = step.steps[k];
+                                                EncodedPolyline points1 = step1.polyline;
+                                                //Decode polyline and add points to list of route coordinates
+                                                List<com.google.maps.model.LatLng> coords = points1.decodePath();
+                                                for (com.google.maps.model.LatLng coord : coords) {

```

Comment:
That is kinda heavy six nested for loops for creating the path, I guess I understand why this can freeze the app if ran on UI Thread.

Reply:
Yeah of course non trivial effort, creating the path and smoothing out is can be an SDK by itself 😄 .